analytic_engine/src/sst/parquet/encoding.rs

```diff
@@ -461,3 +487,457 @@ impl ParquetEncoder {
+    /// type.
+    fn stretch_fixed_length_column(
+        array_ref: &ArrayRef,
+        value_size: usize,
+        value_offsets: &[i32],
+    ) -> Result<ArrayRef> {
+        assert!(!value_offsets.is_empty());
+
+        let values_num = *value_offsets.last().unwrap() as usize;
+        let raw_buffer = array_ref.data().buffers()[0].as_slice();
+        let null_bitmap = array_ref.data().null_bitmap();
+
+        let mut new_buffer = MutableBuffer::new(value_size * values_num);
+        let mut new_null_buffer = MutableBuffer::new(values_num);
+        new_null_buffer = new_null_buffer.with_bitset(values_num, true);
+        let null_slice = new_null_buffer.as_slice_mut();
+
+        let mut length_so_far = 0;
+        for (idx, offset) in (0..raw_buffer.len()).step_by(value_size).enumerate() {
+            let value_num = value_offsets[idx + 1] - value_offsets[idx];
+            if let Some(bitmap) = null_bitmap {
+                if !bitmap.is_set(idx) {
+                    for i in 0..value_num {
+                        bit_util::unset_bit(null_slice, length_so_far + i as usize);
+                    }
+                }
+            }
+            length_so_far += value_num as usize;
+            new_buffer.extend(raw_buffer[offset..offset + value_size].repeat(value_num as usize))
+        }
+        let array_data = ArrayData::builder(array_ref.data_type().clone())
+            .add_buffer(new_buffer.into())
+            .null_bit_buffer(Some(new_null_buffer.into()))
+            .len(values_num)
+            .build()
+            .map_err(|e| Box::new(e) as _)
+            .context(DecodeRecordBatch)?;
+
+        Ok(array_data.into())
+    }
+
+    /// Returns offsets of nested items in ListArray, and populates the
+    /// `values_num`
+    ///
+    /// Note: caller should ensure `array_ref` is ListArray
+    fn get_list_array_offsets(array_ref: &ArrayRef) -> Vec<i32> {
+        let offset_slices = array_ref.data().buffers()[0].as_slice();
+        // *values_num = array_ref.data().child_data()[0].len();
+
+        let mut i32_offsets = Vec::with_capacity(offset_slices.len() / OFFSET_SIZE);
+        for i in (0..offset_slices.len()).step_by(OFFSET_SIZE) {
+            let offset = i32::from_le_bytes(offset_slices[i..i + OFFSET_SIZE].try_into().unwrap());
+            i32_offsets.push(offset);
+        }
+
+        i32_offsets
+    }
+}
+
+impl RecordDecoder for HybridRecordDecoder {
+    /// Decode records from hybrid to columnar format
+    fn decode(&self, arrow_record_batch: ArrowRecordBatch) -> Result<ArrowRecordBatch> {
+        let new_arrow_schema = Self::convert_schema(arrow_record_batch.schema());
+        let arrays = arrow_record_batch.columns();
+        let mut value_offsets = None;
+        // Find value offsets from first `ListArray` column
+        for array_ref in arrays {
+            if matches!(array_ref.data_type(), DataType::List(_)) {
```

Comment:
The same question: what if it is a list array before hybrid encoding? I think we need to record which columns are collapsed rather than expand all those lists. (maybe we don't support creating a nested column for now, but I still think we cannot use `List` as a special array type)